src/couch_views/src/couch_views_reduce_fdb.erl

```diff
@@ -0,0 +1,248 @@
+% Licensed under the Apache License, Version 2.0 (the "License"); you may not
+% use this file except in compliance with the License. You may obtain a copy of
+% the License at
+%
+%   http://www.apache.org/licenses/LICENSE-2.0
+%
+% Unless required by applicable law or agreed to in writing, software
+% distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+% WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+% License for the specific language governing permissions and limitations under
+% the License.
+
+
+-module(couch_views_reduce_fdb).
+
+
+-export([
+    write_doc/5,
+    idx_prefix/3,
+    get_kv_size/3,
+    fold/8
+]).
+
+
+-include("couch_views.hrl").
+-include_lib("couch/include/couch_db.hrl").
+-include_lib("couch_mrview/include/couch_mrview.hrl").
+-include_lib("fabric/include/fabric2.hrl").
+
+
+write_doc(TxDb, Sig, Views, #{deleted := true} = Doc, ExistingViewKeys) ->
+    #{
+        id := DocId
+    } = Doc,
+
+    #{
+        tx := Tx
+    } = TxDb,
+
+    lists:foreach(fun(View) ->
+        #mrview{
+            reduce_funs = ViewReduceFuns,
+            id_num = ViewId
+        } = View,
+
+        ExistingKeys = get_existing_keys(ViewId, ExistingViewKeys),
+        lists:foreach(fun(ViewReduceFun) ->
+            Tree = open_tree(TxDb, Sig, ViewId, ViewReduceFun),
+            delete_keys(Tx, Tree, DocId, ExistingKeys)
+
+        end, ViewReduceFuns)
+
+    end, Views);
+
+write_doc(TxDb, Sig, Views, #{reduce_results := ViewReduceResults, id := DocId},
+    ExistingViewKeys) ->
+    #{
+        tx := Tx
+    } = TxDb,
+
+    lists:foreach(fun({View, ReduceResults}) ->
+        #mrview{
+            reduce_funs = ViewReduceFuns,
+            id_num = ViewId
+        } = View,
+
+        ExistingKeys = get_existing_keys(ViewId, ExistingViewKeys),
+
+        lists:foreach(fun({ViewReduceFun, ReduceResult}) ->
+            Tree = open_tree(TxDb, Sig, ViewId, ViewReduceFun),
+
+            delete_keys(Tx, Tree, DocId, ExistingKeys),
+            add_keys(Tx, Tree, DocId, ReduceResult)
+
+        end, lists:zip(ViewReduceFuns, ReduceResults))
+    end, lists:zip(Views, ViewReduceResults));
+
+write_doc(_TxDb, _Sig, _Views, _Doc, _ExistingViewKeys) ->
+    ok.
+
+
+idx_prefix(DbPrefix, Sig, ViewId) ->
+    Key = {?DB_VIEWS, ?VIEW_DATA, Sig, ?VIEW_REDUCE_RANGE, ViewId},
+    erlfdb_tuple:pack(Key, DbPrefix).
+
+
+get_kv_size(TxDb, Mrst, ViewId) ->
+    #mrst {
+        views = Views,
+        sig = Sig
+    } = Mrst,
+
+    #{
+        tx := Tx
+    } = TxDb,
+
+    [View] = lists:filter(fun(View) -> View#mrview.id_num == ViewId end, Views),
```

Comment:
`View = lists:keyfind(ViewId, #mrview.id_num, Views),` is the idiomatic way to do this.